### PR TITLE
TIP-706-in_group-sorter: Add in group sorter for es

### PR DESCRIPTION
**Pull request by momoss**
Index group and variant group in an array ['in_group']['group_code'].
Add InGroupSorter for variant group product grid, the datagrid send 'in_group_<group_id>' to the PQB, to sort.

Inspire by [InGroupSorter](https://github.com/jjanvier/pim-community-dev/blob/TIP-613-unified/src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/Sorter/InGroupSorter.php)

** ES Sorter Checklist:**
- [x] Add sorter integration tests for the field in the PQB
- [ ] Add missing integration to the PimCatalogXIntegration tests for sort
- [x] Add the sorter + specs
- [ ] Update the reference documentation

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added Behats                      | -
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseSorter.php [hunk 1]
```diff
@@ -55,25 +55,14 @@ public function addFieldSorter($field, $direction, $locale = null, $scope = null
...
-                        "missing" => "_last"
-                    ]
-                ];
-                $this->searchQueryBuilder->addSort($sortClause);
-
-                break;
             case Directions::DESCENDING:
                 $sortClause = [
                     $field => [
-                        "order" => $direction,
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseSorter.php]: simple quotes please (If I merge the metrics PR before yours, I also fixed that up)

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/InGroupSorter.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        parent::addFieldSorter(self::GROUPS_KEY, $direction, $locale, $scope);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsField($field)
+    {
+        return (strpos($field, 'in_group_') !== false);
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/InGroupSorter.php]: in_group_ ?Where do that come from ? can you explain a bit more how variant_group / group are handled in your PR's description ?

**Review by damien-carcel — COMMENTED** [review]
Will be good once reference documentation updated and missing integration tests added.

**Comment by momoss** [comment]
@damien-carcel I think no need more integration test, it sort on a field and there is no custom mapping

**Comment by damien-carcel** [comment]
Yep, that's true, I forgot that. GTM then, just be sure the reference documentation is up to date.

**Review by damien-carcel — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Resources/config/query_builders.yml [hunk 3]
```diff
@@ -120,7 +120,7 @@ services:
     pim_catalog.query.elasticsearch.filter.group:
         class: '%pim_catalog.query.elasticsearch.filter.group.class%'
         arguments:
-            - ['groups']
+            - ['groups', 'variant_group']
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Resources/config/query_builders.yml]: That's wierd, it was not working like this in 1.7 https://github.com/akeneo/pim-community-dev/blob/master/src/Pim/Bundle/CatalogBundle/Resources/config/query_builders.yml#L106

File: src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Sorter/InGroupSorterIntegration.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
+<?php
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Sorter/InGroupSorterIntegration.php]: Add a class or a testcase to show if it works with variant groups  ?

**Review by SamirBoulil — APPROVED** [review]
GTM after fixes / comments answered

File: src/Pim/Bundle/CatalogBundle/Resources/config/query_builders.yml (hunk 3 above)
Comment (momoss) [reply src/Pim/Bundle/CatalogBundle/Resources/config/query_builders.yml]: I'll remove this one , it was a test

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/InGroupSorter.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class InGroupSorter extends BaseSorter implements FieldSorterInterface
+{
+    const GROUPS_KEY = 'groups';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addFieldSorter($field, $direction, $locale = null, $scope = null)
```
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/InGroupSorter.php]: $channel

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/InGroupSorter.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+use Pim\Component\Catalog\Query\Sorter\FieldSorterInterface;
+
+/**
+ * @author    Philippe Mossière <philippe.mossiere@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class InGroupSorter extends BaseFieldSorter implements FieldSorterInterface
+{
+    const GROUPS_KEY = 'groups';
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/InGroupSorter.php]: INLINE PLEASE, ARE YOU SURE YOU WANT TO SUPPORT ANY BC ON THIS CLASS ? IT IS PUBLIC GOD DAMMIT !! Can't stand this damn. :trollface:

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseFieldSorter.php [hunk 7]
```diff
@@ -55,25 +55,14 @@ public function addFieldSorter($field, $direction, $locale = null, $channel = nu
...
-                $this->searchQueryBuilder->addSort($sortClause);
-
-                break;
             case Directions::DESCENDING:
                 $sortClause = [
                     $field => [
-                        'order'   => 'DESC',
-                        'missing' => '_last'
-                    ]
+                        "order"   => $direction,
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseFieldSorter.php]: quote / use 'DESC'

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseFieldSorter.php [hunk 8]
```diff
@@ -55,25 +55,14 @@ public function addFieldSorter($field, $direction, $locale = null, $channel = nu
 
         switch ($direction) {
             case Directions::ASCENDING:
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseFieldSorter.php]: revert the changes on this class please

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseFieldSorter.php (hunk 7 above)
Comment (anaelChardan) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseFieldSorter.php]: Yup, don't understand :)

File: app/phpunit.xml.dist [hunk 9]
```diff
@@ -81,5 +81,9 @@
             <directory suffix="Integration.php">../src/Pim/Bundle/ReferenceDataBundle/tests/integration/PQB</directory>
         </testsuite>
 
+        <testsuite name="PIM_Enrich_Integration_Test">
```
Comment (damien-carcel) [inline app/phpunit.xml.dist]: Please, do not create a new suite for such a small set of tests. Add it to the `PIM_Catalog_PQB_Integration_Test` (and if you could do the same for the ReferenceDataBundle just above, I would be most grateful :smiley:).

File: src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace Pim\Bundle\EnrichBundle\Elasticsearch\Sorter;
+
+use Pim\Bundle\CatalogBundle\Doctrine\ORM\Repository\GroupRepository;
+use Pim\Bundle\CatalogBundle\Elasticsearch\Sorter\Field\BaseFieldSorter;
+use Pim\Component\Catalog\Exception\InvalidArgumentException;
+use Pim\Component\Catalog\Query\Sorter\FieldSorterInterface;
+
+/**
+ * @author    Philippe Mossière <philippe.mossiere@akeneo.com>
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php]: Maybe one thing you could add is a bit of explanation (where is it used and why ?) What do you think ?

**Review by SamirBoulil — APPROVED** [review]
👏  Yep great GTM after comment answered.

File: src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class InGroupSorter extends BaseFieldSorter implements FieldSorterInterface
+{
+    const IN_GROUP_KEY = 'in_group';
+
+    /** @var GroupRepository */
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php]: Interface?

File: src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    protected $groupRepository;
+
+    public function __construct(
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php]: phpdoc

File: src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        GroupRepository $groupRepository,
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php]: Interface?

File: src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        parent::__construct($supportedFields);
+        $this->groupRepository = $groupRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addFieldSorter($field, $direction, $locale = null, $channel = null)
+    {
+        $groupId = str_replace('in_group_', '', $field);
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php:39]: Please use the constant or remove it :)

File: src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml [hunk 15]
```diff
@@ -1,7 +1,9 @@
 parameters:
     pim_enrich.product_query_builder.filter.dummy.class: Pim\Bundle\EnrichBundle\ProductQueryBuilder\Filter\DummyFilter
+    pim_enrich.query.elasticsearch.sorter.in_group.class: Pim\Bundle\EnrichBundle\Elasticsearch\Sorter\InGroupSorter
 
 services:
+    # Filters
```
Comment (damien-carcel) [inline src/Pim/Bundle/EnrichBundle/Resources/config/query_builders.yml:6]: 👍

File: src/Pim/Component/Catalog/Normalizer/Indexing/Product/PropertiesNormalizer.php [hunk 16]
```diff
@@ -19,6 +19,7 @@ class PropertiesNormalizer extends SerializerAwareNormalizer implements Normaliz
 {
     const FIELD_COMPLETENESS = 'completeness';
     const FIELD_IS_ASSOCIATED = 'is_associated';
+    const FIELD_IN_GROUP = 'in_group';
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Normalizer/Indexing/Product/PropertiesNormalizer.php:22]: Duplicated, it is also present in the sorter

**Review by damien-carcel — APPROVED** [review]
GTM once comments fixed.
Please update the `rst` documentation.

File: src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php (hunk 14 above)
Comment (jjanvier) [reply src/Pim/Bundle/EnrichBundle/Elasticsearch/Sorter/InGroupSorter.php:39]: I guess it's related to what to we get in the frontend

**Review by jjanvier — APPROVED** [review]
